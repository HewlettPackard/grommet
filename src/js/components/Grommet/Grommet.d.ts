import * as React from 'react';
import Grommet from '../index';

export declare namespace GrommetTypes {
  type Direction = 'rtl' | 'ltr';
}

export interface GrommetProps extends Grommet.Props {
  dir: GrommetTypes.Direction;
  theme: {};
}

export class Grommet extends React.Component<GrommetProps, undefined> { }

export default Grommet;
