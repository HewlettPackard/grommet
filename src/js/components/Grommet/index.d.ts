import * as React from "react";

export interface GrommetProps {
  full?: boolean;
  theme?: object;
}

declare const Grommet: React.StatelessComponent<GrommetProps>;

export { Grommet };
