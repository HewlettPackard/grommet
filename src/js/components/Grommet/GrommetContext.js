import React from 'react';

// Track which components mounted onto the DOM are coming from
// Grommet (see FocusedContainer.js)
export const GrommetContext = React.createContext(undefined);
