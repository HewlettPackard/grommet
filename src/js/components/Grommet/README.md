## Grommet
This is the top level Grommet container.

[![](https://codesandbox.io/static/img/play-codesandbox.svg)](https://codesandbox.io/s/github/grommet/grommet-site?initialpath=grommet&amp;module=%2Fscreens%2FGrommet.js)
## Usage

```javascript
import { Grommet } from 'grommet';
<Grommet>...</Grommet>
```

## Properties

**dir**

Whether text should be rendered right to left or not. Defaults to
      inherit from the document context.

```
rtl
ltr
```

**theme**

Custom styles for Grommet app component.

```
object
```
  