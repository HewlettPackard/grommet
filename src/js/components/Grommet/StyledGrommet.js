import styled, { css } from 'styled-components';

import { baseStyle } from '../../utils';

const fullStyle = css`
  width: 100vw;
  height: 100vh;
`;

const StyledGrommet = styled.div`
  ${baseStyle}
  ${props => props.full && fullStyle}
  ${props => props.theme.global.font.face}
`;

export default StyledGrommet.extend`
  ${props => props.theme.grommet.extend}
`;
