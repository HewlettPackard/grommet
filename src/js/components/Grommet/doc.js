import { describe, PropTypes } from 'react-desc';

import { getAvailableAtBadge } from '../../utils';

export default (Grommet) => {
  const DocumentedGrommet = describe(Grommet)
    .availableAt(getAvailableAtBadge('Grommet'))
    .description('This is the top level Grommet container.')
    .usage(
      `import { Grommet } from 'grommet';
<Grommet>...</Grommet>`
    );

  DocumentedGrommet.propTypes = {
    dir: PropTypes.oneOf(['rtl', 'ltr']).description(
      `Whether text should be rendered right to left or not. Defaults to
      inherit from the document context.`
    ),
    theme: PropTypes.object.description('Custom styles for Grommet app component.'),
  };

  return DocumentedGrommet;
};
