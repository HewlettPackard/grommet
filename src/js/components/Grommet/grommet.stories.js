import React from 'react';
import { storiesOf } from '@storybook/react';

import { Grommet, Anchor, Box } from 'grommet';
import { Add } from 'grommet-icons';

const customTheme = {
  global: {
    colors: {
      custom: '#cc6633',
    },
  },
};


const Themed = () => (
  <Grommet theme={customTheme}>
    <Box pad='medium'>
      <Anchor icon={<Add />} label='Add' color='custom' />
    </Box>
  </Grommet>
);

storiesOf('Grommet', module)
  .add('Theme', () => <Themed />);
