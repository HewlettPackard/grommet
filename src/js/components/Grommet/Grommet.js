import React, { forwardRef, useEffect, useMemo, useState } from 'react';
import { createGlobalStyle } from 'styled-components';

import {
  ContainerTargetContext,
  ResponsiveContext,
  ThemeContext,
} from '../../contexts';

import {
  deepMerge,
  backgroundIsDark,
  getBreakpoint,
  getDeviceBreakpoint,
  normalizeColor,
} from '../../utils';
import { base as baseTheme } from '../../themes';
import { StyledGrommet } from './StyledGrommet';

const FullGlobalStyle = createGlobalStyle`
  body { margin: 0; }
`;

const deviceResponsive = (userAgent, theme) => {
  // log('--deviceResponsive', userAgent, theme);
  /*
   * Regexes provided for mobile and tablet detection are meant to replace
   * a full-featured specific library due to contributing a considerable size
   * into the bundle.
   *
   * User agents found https://deviceatlas.com/blog/list-of-user-agent-strings
   */
  if (userAgent) {
    if (/(tablet|ipad|playbook|silk)|(android(?!.*mobile))/i.test(userAgent)) {
      return getDeviceBreakpoint('tablet', theme);
    }
    if (/Mobile|iPhone|Android/.test(userAgent)) {
      return getDeviceBreakpoint('phone', theme);
    }
    return getDeviceBreakpoint('computer', theme);
  }
  return undefined;
};

const Grommet = forwardRef((props, ref) => {
  const {
    children,
    full,
    containerTarget = typeof document === 'object' ? document.body : undefined,
    theme: themeProp,
    ...rest
  } = props;

  const { background, dir, themeMode, userAgent } = props;

  const [stateResponsive, setResponsive] = useState();

  const theme = useMemo(() => {
    const nextTheme = deepMerge(baseTheme, themeProp || {});

    const {
      colors: { background: themeBackground },
    } = nextTheme.global;

    nextTheme.dark = (themeMode || nextTheme.defaultMode) === 'dark';
    const color = normalizeColor(background || themeBackground, nextTheme);
    nextTheme.dark = backgroundIsDark(color, nextTheme);
    nextTheme.baseBackground = background || themeBackground;
    // This allows DataTable to intelligently set the background of a pinned
    // header or footer.
    nextTheme.background = nextTheme.baseBackground;

    if (dir) {
      nextTheme.dir = dir;
    }

    return nextTheme;
  }, [background, dir, themeMode, themeProp]);

  useEffect(() => {
    const onResize = () => {
      setResponsive(getBreakpoint(window.innerWidth, theme));
    };
    window.addEventListener('resize', onResize);
    onResize();
    return () => {
      window.removeEventListener('resize', onResize);
    };
  }, [theme]);

  const responsive =
    stateResponsive ||
    deviceResponsive(userAgent, theme) ||
    theme.global.deviceBreakpoints.tablet;

  return (
    <ThemeContext.Provider value={theme}>
      <ResponsiveContext.Provider value={responsive}>
        <ContainerTargetContext.Provider value={containerTarget}>
          <StyledGrommet full={full} {...rest} ref={ref}>
            {children}
          </StyledGrommet>
          {full && <FullGlobalStyle />}
        </ContainerTargetContext.Provider>
      </ResponsiveContext.Provider>
    </ThemeContext.Provider>
  );
});

Grommet.displayName = 'Grommet';

let GrommetDoc;
if (process.env.NODE_ENV !== 'production') {
  // eslint-disable-next-line global-require
  GrommetDoc = require('./doc').doc(Grommet);
}
const GrommetWrapper = GrommetDoc || Grommet;

export { GrommetWrapper as Grommet };
