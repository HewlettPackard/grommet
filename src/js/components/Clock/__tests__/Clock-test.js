import React from 'react';
import renderer from 'react-test-renderer';
import 'jest-styled-components';
import Enzyme, { mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

import { Grommet } from '../../Grommet';
import { Clock } from '../';

Enzyme.configure({ adapter: new Adapter() });

const DURATION = 'PT18H23M34S';
const TIME = 'T18:23:34';
const TIME2 = 'T18:23';
const DATE = '2018-02-22T18:23:34-10:00';

describe('Clock', () => {
  test('time renders', () => {
    const component = renderer.create(
      <Grommet>
        <Clock run={false} type='digital' time={DURATION} />
        <Clock run={false} type='digital' time={TIME} />
        <Clock run={false} type='digital' time={TIME2} />
        <Clock run={false} type='digital' time={DATE} />
      </Grommet>
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('hourLimit renders', () => {
    const component = renderer.create(
      <Grommet>
        <Clock run={false} type='digital' time={DURATION} hourLimit={12} />
        <Clock run={false} type='digital' time={DURATION} hourLimit={24} />
      </Grommet>
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('run renders', (done) => {
    const component = mount(
      <Grommet>
        <Clock type='analog' run='forward' time={DURATION} />
        <Clock type='analog' run='backward' time={DURATION} />
        <Clock type='digital' run='forward' time={DURATION} />
        <Clock type='digital' run='backward' time={DURATION} />
      </Grommet>
    );
    expect(component.getDOMNode()).toMatchSnapshot();

    // give some time for the clock to move and use the callback
    setTimeout(() => {
      expect(component.getDOMNode()).toMatchSnapshot();
      component.unmount();
      done();
    }, 1300);
  });

  ['analog', 'digital'].forEach(type => (
    ['hours', 'minutes', 'seconds'].forEach(precision => (
      ['xsmall', 'small', 'medium', 'large', 'xlarge'].forEach(size => (
        test(`type ${type} precision ${precision} size ${size}`, () => {
          const component = renderer.create(
            <Grommet>
              <Clock
                run={false}
                type={type}
                precision={precision}
                size={size}
                time={DURATION}
              />
            </Grommet>
          );
          const tree = component.toJSON();
          expect(tree).toMatchSnapshot();
        })
      ))
    ))
  ));
});
