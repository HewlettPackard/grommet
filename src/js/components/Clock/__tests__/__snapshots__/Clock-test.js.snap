// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Clock animates second hand 1`] = `
<svg
  class="StyledClock-egFVkc cMDGwi"
  version="1.1"
  width="96"
  height="96"
  preserveAspectRatio="xMidYMid meet"
  viewBox="0 0 96 96"
>
  <circle
    class="StyledClock__StyledCircle-hrpIpF dtvAqv"
    fill="none"
    stroke="#000000"
    cx="48"
    cy="48"
    r="46"
  />
  <line
    class="StyledClock__StyledSecond-bAhDUL CJLmH"
    x1="48"
    y1="48"
    x2="48"
    y2="9"
    stroke="#000000"
    stroke-linecap="round"
  />
  <line
    class="StyledClock__StyledMinute-efBHQO fkJwbu"
    x1="48"
    y1="48"
    x2="48"
    y2="12"
    stroke="#000000"
    stroke-linecap="round"
  />
  <line
    class="StyledClock__StyledHour-lkWeQW cLePsR"
    x1="48"
    y1="48"
    x2="48"
    y2="24"
    stroke="#000000"
    stroke-linecap="round"
  />
</svg>
`;

exports[`Clock animates second hand 2`] = `
<svg
  class="StyledClock-egFVkc cMDGwi"
  version="1.1"
  width="96"
  height="96"
  preserveAspectRatio="xMidYMid meet"
  viewBox="0 0 96 96"
>
  <circle
    class="StyledClock__StyledCircle-hrpIpF dtvAqv"
    fill="none"
    stroke="#000000"
    cx="48"
    cy="48"
    r="46"
  />
  <line
    class="StyledClock__StyledSecond-bAhDUL CJLmH"
    x1="48"
    y1="48"
    x2="48"
    y2="9"
    stroke="#000000"
    stroke-linecap="round"
  />
  <line
    class="StyledClock__StyledMinute-efBHQO fkJwbu"
    x1="48"
    y1="48"
    x2="48"
    y2="12"
    stroke="#000000"
    stroke-linecap="round"
  />
  <line
    class="StyledClock__StyledHour-lkWeQW cLePsR"
    x1="48"
    y1="48"
    x2="48"
    y2="24"
    stroke="#000000"
    stroke-linecap="round"
  />
</svg>
`;

exports[`Clock night renders 1`] = `
<svg
  class="StyledClock-egFVkc cMDGwi"
  version="1.1"
  width="96"
  height="96"
  preserveAspectRatio="xMidYMid meet"
  viewBox="0 0 96 96"
>
  <circle
    class="StyledClock__StyledCircle-hrpIpF grDodx"
    fill="none"
    stroke="#000000"
    cx="48"
    cy="48"
    r="46"
  />
  <line
    class="StyledClock__StyledMinute-efBHQO kArSyA"
    x1="48"
    y1="48"
    x2="48"
    y2="12"
    stroke="#000000"
    stroke-linecap="round"
  />
  <line
    class="StyledClock__StyledHour-lkWeQW ervlRf"
    x1="48"
    y1="48"
    x2="48"
    y2="24"
    stroke="#000000"
    stroke-linecap="round"
  />
</svg>
`;

exports[`Clock night renders 2`] = `
<svg
  class="StyledClock-egFVkc cMDGwi"
  version="1.1"
  width="96"
  height="96"
  preserveAspectRatio="xMidYMid meet"
  viewBox="0 0 96 96"
>
  <circle
    class="StyledClock__StyledCircle-hrpIpF grDodx"
    fill="none"
    stroke="#000000"
    cx="48"
    cy="48"
    r="46"
  />
  <line
    class="StyledClock__StyledMinute-efBHQO kArSyA"
    x1="48"
    y1="48"
    x2="48"
    y2="12"
    stroke="#000000"
    stroke-linecap="round"
  />
  <line
    class="StyledClock__StyledHour-lkWeQW ervlRf"
    x1="48"
    y1="48"
    x2="48"
    y2="24"
    stroke="#000000"
    stroke-linecap="round"
  />
</svg>
`;

exports[`Clock renders 1`] = `
<div
  class="StyledGrommet-gKbdxL fqqdJc"
>
  <svg
    class="StyledClock-egFVkc cMDGwi"
    version="1.1"
    width="96"
    height="96"
    preserveAspectRatio="xMidYMid meet"
    viewBox="0 0 96 96"
  >
    <circle
      class="StyledClock__StyledCircle-hrpIpF grDodx"
      fill="none"
      stroke="#000000"
      cx="48"
      cy="48"
      r="46"
    />
    <line
      class="StyledClock__StyledSecond-bAhDUL CJLmH"
      x1="48"
      y1="48"
      x2="48"
      y2="9"
      stroke="#000000"
      stroke-linecap="round"
    />
    <line
      class="StyledClock__StyledMinute-efBHQO kArSyA"
      x1="48"
      y1="48"
      x2="48"
      y2="12"
      stroke="#000000"
      stroke-linecap="round"
    />
    <line
      class="StyledClock__StyledHour-lkWeQW ervlRf"
      x1="48"
      y1="48"
      x2="48"
      y2="24"
      stroke="#000000"
      stroke-linecap="round"
    />
  </svg>
</div>
`;

exports[`Clock renders 2`] = `
<div
  class="StyledGrommet-gKbdxL fqqdJc"
>
  <svg
    class="StyledClock-egFVkc cMDGwi"
    version="1.1"
    width="96"
    height="96"
    preserveAspectRatio="xMidYMid meet"
    viewBox="0 0 96 96"
  >
    <circle
      class="StyledClock__StyledCircle-hrpIpF grDodx"
      fill="none"
      stroke="#000000"
      cx="48"
      cy="48"
      r="46"
    />
    <line
      class="StyledClock__StyledSecond-bAhDUL TwNiW"
      x1="48"
      y1="48"
      x2="48"
      y2="9"
      stroke="#000000"
      stroke-linecap="round"
    />
    <line
      class="StyledClock__StyledMinute-efBHQO dZbggh"
      x1="48"
      y1="48"
      x2="48"
      y2="12"
      stroke="#000000"
      stroke-linecap="round"
    />
    <line
      class="StyledClock__StyledHour-lkWeQW gQoCKI"
      x1="48"
      y1="48"
      x2="48"
      y2="24"
      stroke="#000000"
      stroke-linecap="round"
    />
  </svg>
</div>
`;
